File: year2019/day14.py
# 9 ORE => 2 A
# 8 ORE => 3 B
# 7 ORE => 5 C
# 3 A, 4 B => 1 AB
# 5 B, 7 C => 1 BC
# 4 C, 1 A => 1 CA
# 2 AB, 3 BC, 4 CA => 1 FUEL
#
# 4CA = 4*4 C + 4*1A
# 3BC = 15B + 21C
# 2AB = 6A + 8B
#
# total = 10A + 23B + 37C
#
# 10 = 4*9 = 45
# 8*8 = 64
# 40/5 = 56
#
# 45+64+56 = 165

with open("day14.txt", 'r') as fp:
    chemical_dict = {}
    for line in fp.readlines():
        input,output = line.strip().split(' => ')

# TODO: Turn this into queue. Start with fuel, enqueue everything, pop off queue then add next ingredient, repeat until queue contains only ore, sum ore = win

File: day14.jl
Pkg.add("DataStructures");
using DataStructures;

chem_dict = Dict()
f = open("day14.txt");
for line in readlines(f)
    input,output = split(line, " => ")
    o_quant, o_name = split(output)
    push!(chem_dict,o_name => Dict("quantity"=>o_quant, "required_mats"=>[(parse(Int64,each[1]),each[2]) for each in [split(mat) for mat in split(input,", ")]]))
end

function compute_ore_required(fuel_amount)
    global chem_dict
    production_queue = Queue{Any}()
    reserves = DefaultDict(0)
    enqueue!(production_queue,(fuel_amount,"FUEL"))
    while isempty(production_queue) != true
        chem = dequeue!(production_queue)
        quantity,name = chem[1],chem[2]
        if name == "ORE" # Terminal. We are now adding ORES
            reserves["ORE"] += quantity
        elseif reserves[name] >= quantity # We have enough w.o producing
            reserves[name] -= quantity
        else
            required = quantity - reserves[name] #amount we need
            runs = ceil(required/parse(Int64,chem_dict[name]["quantity"])) # Each run produces no fractional output. Only exacts
            reserves[name] = runs * parse(Int64,chem_dict[name]["quantity"]) - required # Compute how much new stuff we got subtract required for current rxn
            for each in chem_dict[name]["required_mats"]
                req_quant,req_name = each[1],each[2]
                enqueue!(production_queue,(req_quant*runs,req_name)) # multiply each ele in mats by how many runs of production we need and enqueue
            end
        end
    end
    return reserves["ORE"]
end

println(compute_ore_required(1))
# Binary search
low = 0
high = 10^12
while low+1 < high
    global low,high
    mid = low + floor((high - low)//2)
    result =  compute_ore_required(mid)
    if result < 10^12
        low = mid
    else
        high = mid - 1
    end
end
println(Int(low))
